Mathlib/Data/List/Basic.lean

```diff
@@ -2559,6 +2567,69 @@ theorem foldlRecOn_nil {C : β → Sort*} (op : β → α → β) (b) (hb : C b)
   rfl
 #align list.foldl_rec_on_nil List.foldlRecOn_nil
 
+private lemma append_singleton_append_mem_left_of_length_lt {x₁ x₂ z₁ z₂ : List α} {Y₁ Y₂ : α}
+    (together : x₁ ++ [Y₁] ++ z₁ = x₂ ++ [Y₂] ++ z₂) (longer : x₂.length < x₁.length) :
+    Y₂ ∈ x₁ := by
+  have middle := congr_fun (congr_arg get? together) x₂.length
+  rw [
+    append_assoc x₂,
+    get?_append_right x₂.length.le_refl,
+    Nat.sub_self,
+    singleton_append,
+    get?_cons_zero,
+    append_assoc x₁,
+    get?_append longer,
+  ] at middle
+  exact get?_mem middle
+
+private lemma ge_of_add_eq_add_of_le {a b c d : ℕ} (total : a + b = c + d) (a_lt_c : a < c) :
+    d < b := by
+  by_contra contr
+  rw [not_lt] at contr
+  exact (Nat.ne_of_lt (add_lt_add_of_lt_of_le a_lt_c contr)) total
```

Comment:
```suggestion
  exact (add_lt_add_of_lt_of_le a_lt_c contr).ne total
```

Reply:
Leading to the question:
* can you state this more generally than just for `Nat`?
* does mathlib already have that statement?

Reply:
See new version.